aten/src/ATen/core/boxing/KernelFunction.h

```diff
@@ -0,0 +1,466 @@
+#pragma once
+
+#include <ATen/core/stack.h>
+#include <c10/util/TypeList.h>
+#include <ATen/core/boxing/kernel_functor.h>
+#include <ATen/core/boxing/kernel_function.h>
+#include <ATen/core/boxing/kernel_lambda.h>
+
+namespace c10 {
+
+namespace detail {
+template<class Type>
+constexpr uint64_t hashType() {
+  return typeid(Type).hash_code()
+      + 1000 * std::is_lvalue_reference<Type>::value
+      + 5000 * std::is_rvalue_reference<Type>::value
```

Comment:
value is true or false right? There's no reason to have such a large step here?

Reply:
I use larger steps because I can't assume anything about the distribution of typeids. It's very possible that compilers choose adjacent type ids for types, i.e. if one type has id `250` then another type will have id `251` and so on. So I can't just use the numbers adjacent to the type id for representing const, that would be a collision. If I go larger steps away, it is less likely to collide. It's still not perfect, but less likely.